File: flink-demo/flinkScalaCode/src/main/scala/com/yw/flink/example/scalacases/case14_tableAndStreamAndSql/Case08_FromChangelogStream.scala
package com.yw.flink.example.scalacases.case14_tableAndStreamAndSql

import org.apache.flink.api.common.typeinfo.Types
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.table.api.{Schema, Table}
import org.apache.flink.table.api.bridge.scala.StreamTableEnvironment
import org.apache.flink.table.connector.ChangelogMode
import org.apache.flink.types._

/**
  * Flink DataStream转换成Table - tableEnvironment.fromChangelogStream(...)
  */
object Case08_FromChangelogStream {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment

    env.setParallelism(1)
    val tableEnv: StreamTableEnvironment = StreamTableEnvironment.create(env)

    val ds = env.fromElements(
      Row.ofKind(RowKind.INSERT, "zs", Int.box(18)),
      Row.ofKind(RowKind.INSERT, "ls", Int.box(19)),
      Row.ofKind(RowKind.INSERT, "ww", Int.box(20)),
      Row.ofKind(RowKind.UPDATE_BEFORE, "zs", Int.box(18)),
      Row.ofKind(RowKind.UPDATE_AFTER, "zs", Int.box(21)),
      Row.ofKind(RowKind.DELETE, "ls", Int.box(19)),
      Row.ofKind(RowKind.INSERT, "xx", Int.box(108)),
    )(Types.ROW(Types.STRING, Types.INT))

    val result: Table = tableEnv.fromChangelogStream(
      ds,
      Schema.newBuilder()
        .primaryKey("f0")
        .build(),
      ChangelogMode.all()
    )

    result.printSchema();
    result.execute().print();


  }

}

File: flink-demo/flinkJavaCode/src/main/java/com/yw/flink/example/javacases/case14_flinksql/Case08_FromChangelogStream.java
package com.yw.flink.example.javacases.case14_flinksql;

import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.api.Schema;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
import org.apache.flink.table.connector.ChangelogMode;
import org.apache.flink.types.Row;
import org.apache.flink.types.RowKind;

/**
 * Flink DataStream转换成Table - tableEnvironment.fromChangelogStream(...)
 */
public class Case08_FromChangelogStream {
    public static void main(String[] args) {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        //创建TableEnvironment
        StreamTableEnvironment tableEnv = StreamTableEnvironment.create(env);

        //设置并行度为1
        env.setParallelism(1);

        //准备数据流
        DataStreamSource<Row> dataStream = env.fromElements(
                Row.ofKind(RowKind.INSERT, "zs", 18),
                Row.ofKind(RowKind.INSERT, "ls", 19),
                Row.ofKind(RowKind.INSERT, "ww", 20),
                Row.ofKind(RowKind.UPDATE_BEFORE, "zs", 18),
                Row.ofKind(RowKind.UPDATE_AFTER, "zs", 21),
                Row.ofKind(RowKind.DELETE, "ls", 19),
                Row.ofKind(RowKind.INSERT, "xx", 100)
        );

        Table result = tableEnv.fromChangelogStream(
                dataStream,
                Schema.newBuilder()
                        .primaryKey("f0")
                        .build(),
                ChangelogMode.all()
        );

        result.printSchema();
        result.execute().print();
    }
}
